inc/class-core-sitemaps-provider.php

```diff
@@ -194,15 +216,136 @@ public function get_sitemaps() {
 			}
 
 			$total = $this->max_num_pages( $name );
-			for ( $i = 1; $i <= $total; $i ++ ) {
-				$slug       = implode( '-', array_filter( array( $this->slug, $name, (string) $i ) ) );
-				$sitemaps[] = $slug;
+
+			for ( $page = 1; $page <= $total; $page ++ ) {
+				$loc        = $this->get_sitemap_url( $name, $page );
+				$lastmod    = $this->get_sitemap_lastmod( $name, $page );
+				$sitemaps[] = array(
+					'loc'     => $loc,
+					'lastmod' => $lastmod,
+				);
 			}
 		}
 
 		return $sitemaps;
 	}
 
+	/**
+	 * Get the URL of a sitemap entry.
+	 *
+	 * @param string $name The name of the sitemap.
+	 * @param int    $page The page of the sitemap.
+	 * @return string The composed URL for a sitemap entry.
+	 */
+	public function get_sitemap_url( $name, $page ) {
+		global $wp_rewrite;
+
+		$basename = sprintf(
+			'/sitemap-%1$s.xml',
+			// Accounts for cases where name is not included, ex: sitemaps-users-1.xml.
+			implode( '-', array_filter( array( $this->slug, $name, (string) $page ) ) )
+		);
+
+		$url = home_url( $basename );
+
+		if ( ! $wp_rewrite->using_permalinks() ) {
+			$url = add_query_arg(
+				array(
+					'sitemap'  => $this->slug,
+					'sub_type' => $name,
+					'paged'    => $page,
+				),
+				home_url( '/' )
+			);
+		}
+
+		return $url;
+	}
+
+	/**
+	 * Get the last modified date for a sitemap page.
+	 *
+	 * This will be overridden in provider subclasses.
+	 *
+	 * @param string $name The name of the sitemap.
+	 * @param int    $page The page of the sitemap being returned.
+	 * @return string The GMT date of the most recently changed date.
+	 */
+	public function get_sitemap_lastmod( $name, $page ) {
+		$type = implode( '_', array_filter( array( $this->slug, $name, (string) $page ) ) );
+
+		// Check for an option.
+		$lastmod = get_option( "core_sitemaps_lasmod_$type", '' );
+
+		// If blank, schedule a job.
+		if ( empty( $lastmod ) && ! wp_doing_cron() ) {
+			$event_args = array( $this->slug, $name, $page );
+
+			// Don't schedule a duplicate job.
+			if ( ! wp_next_scheduled( 'core_sitemaps_calculate_lastmod', $event_args ) ) {
+				wp_schedule_single_event( time() + 500, 'core_sitemaps_calculate_lastmod', $event_args );
+			}
+		}
+
+		return $lastmod;
+	}
+
+	/**
+	 * Calculate lastmod date for a sitemap page.
+	 *
+	 * Calculated value is saved to the database as an option.
+	 *
+	 * @param string $type    The object type of the page: posts, taxonomies, users, etc.
+	 * @param string $subtype The object subtype if applicable, e.g., post type, taxonomy type.
+	 * @param int    $page    The page number.
+	 */
+	public function calculate_sitemap_lastmod( $type, $subtype, $page ) {
```

Comment:
I've opened #91 to address confusing naming of parameters and properties that I'd like to see addressed, but is outside of the scope of this issue.